core/src/main/java/tc/oc/pgm/channels/ChannelManager.java

```diff
@@ -0,0 +1,264 @@
+    Component reason =
+        muteReason.isPresent() ? text(muteReason.get()) : translatable("moderation.mute.noReason");
+
+    throw exception("moderation.mute.message", reason.color(NamedTextColor.AQUA));
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public <T> void sendMessage(ChannelMessageEvent<T> event) {
+    if (event.isCancelled()) {
+      if (event.getSender() != null && event.getCancellationReason() != null)
+        event.getSender().sendWarning(event.getCancellationReason());
+      return;
+    }
+
+    Component finalMessage =
+        event
+            .getChannel()
+            .formatMessage(event.getTarget(), event.getSender(), text(event.getMessage()));
+    Identity senderId = identity(event.getSender().getId());
+    event.getViewers().forEach(player -> player.sendMessage(senderId, finalMessage));
+  }
+
+  @EventHandler
+  public void onPlayerTabComplete(PlayerChatTabCompleteEvent event) {
+    if (event.getChatMessage().trim().equals(event.getLastToken())) {
+      char first = event.getLastToken().charAt(0);
+      if (shortcuts.get(first) != null) {
+        List<String> suggestions =
+            Players.getPlayerNames(event.getPlayer(), event.getLastToken().substring(1));
+        suggestions.replaceAll(s -> first + s);
+
+        event.getTabCompletions().addAll(suggestions);
+      }
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void onPlayerJoin(final PlayerJoinEvent event) {
+    MatchPlayer player = PGM.get().getMatchManager().getPlayer(event.getPlayer());
+    if (player == null) return;
+    selectedChannel.put(
+        player.getBukkit(), findChannelBySetting(player.getSettings().getValue(SettingKey.CHAT)));
+  }
+
+  @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
+  public void onChat(AsyncPlayerChatEvent event) {
+    if (CHAT_EVENT_CACHE.getIfPresent(event) == null) {
+      event.setCancelled(true);
+    } else {
+      CHAT_EVENT_CACHE.invalidate(event);
+      return;
+    }
+
+    final MatchPlayer player = PGM.get().getMatchManager().getPlayer(event.getPlayer());
+    if (player == null) return;
+
+    final String message = event.getMessage().trim();
+    try {
+      processChat(player, message);
+    } catch (TextException e) {
+      // Allow sub-handlers to throw command exceptions just fine
+      player.sendWarning(e);
+    }
+  }
+
+  private Channel<?> findChannelBySetting(SettingValue setting) {
+    for (Channel<?> channel : channels) if (setting == channel.getSetting()) return channel;
+
+    return globalChannel;
+  }
+
+  public void setChannel(MatchPlayer player, SettingValue value) {
+    selectedChannel.put(player.getBukkit(), findChannelBySetting(value));
+  }
+
+  public void setChannel(MatchPlayer player, Channel<?> channel) {
+    Channel<?> previous = selectedChannel.get(player.getBukkit());
+    selectedChannel.put(player.getBukkit(), channel);
```

Comment:
```suggestion
    Channel<?> previous = selectedChannel.put(player.getBukkit(), channel);
```
Also, consider using UUIDs for maps instead of players or match players. Avoids mem leaks from holding onto entities you shouldn't